cmd/transcriber/ogg/writer.go

```diff
@@ -0,0 +1,254 @@
+	data := i.createPage(oggIDHeader, pageHeaderTypeBeginningOfStream, 0, i.pageIndex)
+	if err := i.writeToStream(data); err != nil {
+		return err
+	}
+	i.pageIndex++
+
+	// Comment Header
+	oggCommentHeader := make([]byte, 21)
+	copy(oggCommentHeader[0:], commentPageSignature)        // Magic Signature 'OpusTags'
+	binary.LittleEndian.PutUint32(oggCommentHeader[8:], 5)  // Vendor Length
+	copy(oggCommentHeader[12:], "pion")                     // Vendor name 'pion'
+	binary.LittleEndian.PutUint32(oggCommentHeader[17:], 0) // User Comment List Length
+
+	// RFC specifies that the page where the CommentHeader completes should have a granule position of 0
+	data = i.createPage(oggCommentHeader, pageHeaderTypeContinuationOfStream, 0, i.pageIndex)
+	if err := i.writeToStream(data); err != nil {
+		return err
+	}
+	i.pageIndex++
+
+	return nil
+}
+
+const (
+	pageHeaderSize = 27
+)
+
+func (i *Writer) createPage(payload []uint8, headerType uint8, granulePos uint64, pageIndex uint32) []byte {
+	i.lastPayloadSize = len(payload)
+	nSegments := (len(payload) / 255) + 1 // A segment can be at most 255 bytes long.
+
+	page := make([]byte, pageHeaderSize+i.lastPayloadSize+nSegments)
+
+	copy(page[0:], pageHeaderSignature)                 // page headers starts with 'OggS'
+	page[4] = 0                                         // Version
+	page[5] = headerType                                // 1 = continuation, 2 = beginning of stream, 4 = end of stream
+	binary.LittleEndian.PutUint64(page[6:], granulePos) // granule position
+	binary.LittleEndian.PutUint32(page[14:], i.serial)  // Bitstream serial number
+	binary.LittleEndian.PutUint32(page[18:], pageIndex) // Page sequence number
+	page[26] = uint8(nSegments)                         // Number of segments in page.
+
+	// Filling segment table with the lacing values.
+	// First (nSegments - 1) values will always be 255.
+	for i := 0; i < nSegments-1; i++ {
+		page[pageHeaderSize+i] = 255
+	}
+	// The last value will be the remainder.
+	page[pageHeaderSize+nSegments-1] = uint8(len(payload) % 255)
+
+	copy(page[pageHeaderSize+nSegments:], payload) // Payload goes after the segment table, so at pageHeaderSize+nSegments.
+
+	var checksum uint32
+	for index := range page {
+		checksum = (checksum << 8) ^ i.checksumTable[byte(checksum>>24)^page[index]]
+	}
+
+	binary.LittleEndian.PutUint32(page[22:], checksum) // Checksum - generating for page data and inserting at 22th position into 32 bits
+
+	return page
+}
+
+// WriteRTP adds a new packet and writes the appropriate headers for it
+func (i *Writer) WriteRTP(packet *rtp.Packet, gap uint64) error {
+	if packet == nil {
+		return errInvalidNilPacket
+	}
+	if len(packet.Payload) == 0 {
+		return nil
+	}
+
+	opusPacket := codecs.OpusPacket{}
+	if _, err := opusPacket.Unmarshal(packet.Payload); err != nil {
+		// Only handle Opus packets
+		return err
+	}
+
+	payload := opusPacket.Payload[0:]
```

Comment:
I don't understand what this part is doing? Looking at the implementation of `OpusPacket`, `opusPacket.Unmarshal(packet.Payload)` just assigns `packet.Payload` to `opusPacket.Payload`... and then we assign that to `payload`. How is that different from just using `packet.Payload` directly?

Reply:
Need to double check since I didn't write that code which is ported from https://github.com/pion/webrtc/tree/master/pkg/media as I needed to fix a bug and wasn't sure I'd get it upstream soon enough.

Reply:
At a first glance I think they are just playing it safe by following the interface. In the case of Opus unmarshalling isn't doing much other than some sanity checks on the buffer but in general (other codecs) that may not always be the case. It can be simplified though.